### PR TITLE
PlayingstateUI (heet nu UI ipv GUI)

### PR DESCRIPTION
**Pull request by nessemis**

### CODE REVIEW

File: MeesGame/Gameplay/GameManagement/GameEnvironment.cs [hunk 1]
```diff
@@ -16,6 +16,7 @@ public class GameEnvironment : Game
     protected static Random random;
     protected static AssetManager assetManager;
     protected static GameSettingsManager gameSettingsManager;
+    protected static GameEnvironment gameEnvirontment;
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/GameManagement/GameEnvironment.cs]: The word is Environment, without a T. ctrl+R ctrl + R

File: MeesGame/Gameplay/GameManagement/GameEnvironment.cs [hunk 2]
```diff
@@ -55,6 +57,11 @@ public static GameSettingsManager GameSettingsManager
         get { return gameSettingsManager; }
     }
 
+    public static GameEnvironment GetGameEnvironment
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/GameManagement/GameEnvironment.cs]: The property should be called GameEnvironment

File: MeesGame/Gameplay/GameObjects/PlayerState.cs [hunk 3]
```diff
@@ -19,6 +19,9 @@ public InventoryItemType type
 
     class Inventory
     {
+        public delegate void InventoryChangedEventHandler();
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/GameObjects/PlayerState.cs]: Unused

File: MeesGame/Gameplay/GameObjects/PlayerState.cs [hunk 4]
```diff
@@ -28,6 +31,11 @@ public Inventory()
         {
             Items = new List<InventoryItem>();
         }
+
+        private void InventoryChanged()
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/GameObjects/PlayerState.cs]: Please remove this function, it has too little functionality

File: MeesGame/Gameplay/GameStates/LoadMenuState.cs [hunk 5]
```diff
@@ -8,43 +8,43 @@ namespace MeesGame
     internal class LoadMenuState : IGameLoopObject
     {
         //we need to use a container, because only elements in a container can eat the input of the other elements
-        private UIContainer uiContainer;
+        private UIContainer UIContainer;
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/GameStates/LoadMenuState.cs]: should be uiContainer

File: MeesGame/Gameplay/GameStates/LoadMenuState.cs [hunk 6]
```diff
@@ -8,43 +8,43 @@ namespace MeesGame
...
 
         private FileExplorer levelExplorer;
         private FileExplorer aiExplorer;
         private Button startButton;
 
-        public LoadMenuState(ContentManager content)
+        public LoadMenuState()
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/GameStates/LoadMenuState.cs:16]: Much better

File: MeesGame/Gameplay/UIObjects/UIObject.cs [hunk 7]
```diff
@@ -25,6 +35,114 @@ public abstract class UIObject : IGameLoopObject
...
+        /// <summary>
+        /// UI elements are able to override this at will, for example if the UI changes over time
+        /// </summary>
+        /// <param name="gameTime">current time</param>
+        public virtual void Update(GameTime gameTime)
+        {
+        }
+
+        /// <summary>
+        /// draws the UIel
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/UIObject.cs]: Not helpful

File: MeesGame/Gameplay/UIObjects/UIObject.cs [hunk 8]
```diff
@@ -40,11 +158,13 @@ public bool Clicked
...
+        public virtual bool Invalidate
         {
-            get { return invalidated; }
-            set {invalidated = value;
-                if(parent != null)
+            get { return needsToBeInvalidated; }
+            set
+            {
+                needsToBeInvalidated = value;
+                if (parent != null)
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/UIObject.cs]: This condition should be 
    if(parent!=null && value)

File: MeesGame/Gameplay/UIObjects/UIObject.cs [hunk 9]
```diff
@@ -40,11 +158,13 @@ public bool Clicked
             get { return clicked; }
         }
 
-        public bool Invalidate
+        public virtual bool Invalidate
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/UIObject.cs]: Please make this a function Invalidate(). If it is not a function it should not be a verb, it should be Invalidated.
Also, add documentation to this property.

File: MeesGame/Gameplay/UIObjects/UIObject.cs [hunk 10]
```diff
@@ -112,91 +232,12 @@ public UIContainer Parent
...
-        /// <param name="spriteBatch"></param>
-        public void Draw(GameTime gameTime, SpriteBatch spriteBatch)
-        {
-            if (Invalidate == true)
-            {
-                renderTarget?.Dispose();
-                textureRenderer.Render(gameTime, spriteBatch.GraphicsDevice, DrawTask, dimensions, out renderTarget);
-                invalidated = false;
+            get { return visible; }
+            set { visible = value;
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/UIObject.cs]: Please add a newline here

File: MeesGame/Gameplay/UIObjects/UIObject.cs (hunk 9 above)
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/UIObject.cs]: Please make this a function Invalidate(). If it is not a function it should not be a verb, it should be Invalidated.
Also, add documentation to this property.

File: MeesGame/Gameplay/UIObjects/UIObject.cs [hunk 11]
```diff
@@ -25,6 +35,114 @@ public abstract class UIObject : IGameLoopObject
...
+        /// </summary>
+        /// <param name="gameTime"></param>
+        /// <param name="spriteBatch"></param>
+        public void Draw(GameTime gameTime, SpriteBatch spriteBatch)
+        {
+            if (!visible) return;
+
+            if (BackgroundColor != null)
+                spriteBatch.Draw(SolidWhiteTexture(spriteBatch.GraphicsDevice), RelativeRectangle, (Color) BackgroundColor);
+            if (Invalidate == true)
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/UIObject.cs]: if (Invalidated)

File: MeesGame/Gameplay/UIObjects/UIObject.cs [hunk 12]
```diff
@@ -5,18 +5,28 @@ namespace MeesGame
...
+        public delegate void ClickEventHandler(UIObject UIObject);
+        public event ClickEventHandler OnClick;
+
+        /// <summary>
+        /// Texture that holds a solid white color. It can be used to draw solid backgrounds
+        /// </summary>
+        private static Texture2D solidWhiteTexture;
+
         /// <summary>
         /// location = relative location to the parent of the object
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/UIObject.cs]: Please use the default format for documentation

File: MeesGame/Gameplay/UIObjects/ScrollBar.cs [hunk 13]
```diff
@@ -5,29 +5,102 @@ namespace MeesGame
...
-
         //Variables defining the colors of the scrollbar
-        private readonly Color barColor = Color.Aqua;
-        private readonly Color backColor = Color.Beige;
+        private readonly Color scrollbarColor = Color.Aqua;
 
         //Stores the total length of elements in the container.
-        private int totalElementsSize;
+        private int parentHeightWhenShowingAllChildren;
         //The scroll distance before we began dragging. This variable is needed because otherwise we'd have
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/ScrollBar.cs]: Please use documention tags

File: MeesGame/Gameplay/UIObjects/Objects/Button.cs [hunk 14]
```diff
@@ -79,17 +74,26 @@ public override void DrawTask(GameTime gameTime, SpriteBatch spriteBatch)
...
+        {
+            get
             {
-                OnClick?.Invoke(this);
+                return base.Invalidate;
             }
 
+            set
+            {
+                base.Invalidate = true;
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/Objects/Button.cs]: What is this????? Are you sure Invalidate = false should set Invalidate to true?

File: MeesGame/Gameplay/UIObjects/Objects/Button.cs [hunk 15]
```diff
@@ -27,12 +24,6 @@ public class Button : UIObject
...
         protected bool selected = false;
 
-        public bool Selected
-        {
-            get { return selected; }
-            set { selected = value; }
-        }
-
         /// <summary>
         /// Method used to create a customizable button
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/Objects/Button.cs:31]: You can reword this: Creates a customizable button

File: MeesGame/Gameplay/UIObjects/Objects/Button.cs [hunk 16]
```diff
@@ -6,9 +6,6 @@ namespace MeesGame
 {
     public class Button : UIObject
     {
-        public delegate void ClickEventHandler(Button button);
-        public event ClickEventHandler OnClick;
-
         /// <summary>
         /// Text = the text the button displays
```
Comment (maxmeijer42) [inline MeesGame/Gameplay/UIObjects/Objects/Button.cs:11]: Use the docs format

**Review by maxmeijer42 — CHANGES_REQUESTED** [review]
There are some naming, documentation and convention issues.

File: MeesGame/Gameplay/GameManagement/GameEnvironment.cs (hunk 2 above)
Comment (jensj12) [reply MeesGame/Gameplay/GameManagement/GameEnvironment.cs]: GameEnvironment generates an error, if you have better suggestions, please let us know.

**Comment by jensj12** [comment]
Remember to not put multiple things in one PR. This contains both an in-game overlay and extensions to the level editor. It is much easier to merge smaller PR's.

File: MeesGame/Gameplay/GameManagement/GameEnvironment.cs (hunk 2 above)
Comment (maxmeijer42) [reply MeesGame/Gameplay/GameManagement/GameEnvironment.cs]: The normal way to handle singletons is to call it Instance. Then you can use GameEnvironment.Instance. So I suggest renaming it to Instance

**Comment by nessemis** [comment]
@maxmeijer42 New commits that have adressed the requested changes (except the one of Jens, because I couldn't undo all the changes)

**Review by maxmeijer42 — APPROVED** [review]
Going to approve this now, we'll refactor later.

**Pull request merged**